### PR TITLE
[feature](pipelline)add local merge sort exchaner

### PR DESCRIPTION
**Pull request by Mryange**
## Proposed changes
```
                      DATA_STREAM_SINK_OPERATOR  (id=2,dst_id=2):
                            -  BlocksProduced:  sum  2,  avg  1,  max  2,  min  0
                            -  CloseTime:  avg  50.352us,  max  59.404us,  min  41.300us
                            -  ExecTime:  avg  357.197us,  max  428.119us,  min  286.275us
                            -  InitTime:  avg  103.853us,  max  104.291us,  min  103.416us
                            -  InputRows:  sum  60,  avg  30,  max  60,  min  0
                            -  MemoryUsage:  sum  ,  avg  ,  max  ,  min  
                                -  PeakMemoryUsage:  sum  0.00  ,  avg  0.00  ,  max  0.00  ,  min  0.00  
                            -  OpenTime:  avg  128.654us,  max  163.939us,  min  93.370us
                            -  RowsProduced:  sum  60,  avg  30,  max  60,  min  0
                            -  WaitForDependencyTime:  avg  0ns,  max  0ns,  min  0ns
                                -  WaitForRpcBufferQueue:  avg  0ns,  max  0ns,  min  0ns
                          LOCAL_EXCHANGE_OPERATOR  (LOCAL_MERGE_SORT)  (id=-3):
                                -  BlocksProduced:  sum  2,  avg  1,  max  2,  min  0
                                -  CloseTime:  avg  0ns,  max  0ns,  min  0ns
                                -  ExecTime:  avg  873.747us,  max  1.737ms,  min  9.663us
                                -  GetBlockFailedTime:  sum  0,  avg  0,  max  0,  min  0
                                -  InitTime:  avg  466ns,  max  495ns,  min  437ns
                                -  MemoryUsage:  sum  ,  avg  ,  max  ,  min  
                                    -  PeakMemoryUsage:  sum  0.00  ,  avg  0.00  ,  max  0.00  ,  min  0.00  
                                -  OpenTime:  avg  19.531us,  max  31.460us,  min  7.602us
                                -  ProjectionTime:  avg  0ns,  max  0ns,  min  0ns
                                -  RowsProduced:  sum  60,  avg  30,  max  60,  min  0
                                -  WaitForDependency[LOCAL_EXCHANGE_OPERATOR_DEPENDENCY]Time:  avg  14.130ms,  max  14.701ms,  min  13.558ms
                  Pipeline  :  1(instance_num=2):
                      LOCAL_EXCHANGE_SINK_OPERATOR  (LOCAL_MERGE_SORT)  (id=-3):
                            -  CloseTime:  avg  0ns,  max  0ns,  min  0ns
                            -  ExecTime:  avg  59.515us,  max  68.50us,  min  50.981us
                            -  InitTime:  avg  8.415us,  max  8.695us,  min  8.135us
                            -  InputRows:  sum  60,  avg  30,  max  30,  min  30
                            -  MemoryUsage:  sum  ,  avg  ,  max  ,  min  
                                -  PeakMemoryUsage:  sum  0.00  ,  avg  0.00  ,  max  0.00  ,  min  0.00  
                            -  OpenTime:  avg  1.498us,  max  1.630us,  min  1.366us
                            -  WaitForDependency[LOCAL_EXCHANGE_SINK_DEPENDENCY]Time:  avg  0ns,  max  0ns,  min  0ns
                          SORT_OPERATOR  (id=1  ,  nereids_id=108):
                                -  PlanInfo
                                      -  order  by:  s_suppkey  ASC
                                      -  TOPN  OPT
                                      -  offset:  0
                                      -  limit:  30
                                -  BlocksProduced:  sum  2,  avg  1,  max  1,  min  1
                                -  CloseTime:  avg  0ns,  max  0ns,  min  0ns
                                -  ExecTime:  avg  10.593us,  max  18.22us,  min  3.165us
                                -  InitTime:  avg  0ns,  max  0ns,  min  0ns
                                -  MemoryUsage:  sum  ,  avg  ,  max  ,  min  
                                    -  PeakMemoryUsage:  sum  0.00  ,  avg  0.00  ,  max  0.00  ,  min  0.00  
                                -  OpenTime:  avg  0ns,  max  0ns,  min  0ns
                                -  ProjectionTime:  avg  0ns,  max  0ns,  min  0ns
```

### CODE REVIEW

**Comment by Mryange** [comment]
run buildall

File: be/src/pipeline/exec/sort_sink_operator.cpp [hunk 1]
```diff
@@ -185,4 +186,18 @@ void SortSinkOperatorX::reset(RuntimeState* state) {
...
     local_state._shared_state->sorter->reset();
 }
+
+Status SortSinkOperatorX::build_merger(RuntimeState* state,
+                                       std::unique_ptr<vectorized::VSortedRunMerger>& merger,
+                                       RuntimeProfile* profile) {
+    vectorized::VSortExecExprs vsort_exec_exprs;
+    RETURN_IF_ERROR(vsort_exec_exprs.init(_sort_info, state->obj_pool()));
+    RETURN_IF_ERROR(vsort_exec_exprs.prepare(state, _child_x->row_desc(), _row_descriptor));
+    RETURN_IF_ERROR(vsort_exec_exprs.open(state));
```
Comment (Gabriel39) [inline be/src/pipeline/exec/sort_sink_operator.cpp:196]: `open` should be called in local state 's open function

File: be/src/pipeline/exec/sort_source_operator.h [hunk 2]
```diff
@@ -46,10 +48,16 @@ class SortSourceOperatorX final : public OperatorX<SortLocalState> {
...
+    bool merge_by_exchange() const { return _merge_by_exchange; }
     const vectorized::SortDescription& get_sort_description(RuntimeState* state) const;
 
+    void set_sort_sink(SortSinkOperatorX* sort_sink) { _sort_sink = sort_sink; }
+    SortSinkOperatorX* sort_sink() { return _sort_sink; }
+
 private:
     friend class SortLocalState;
+    const bool _merge_by_exchange;
+    SortSinkOperatorX* _sort_sink = nullptr;
```
Comment (Gabriel39) [inline be/src/pipeline/exec/sort_source_operator.h:60]: we should not keep a `_sort_sink` here

File: be/src/pipeline/local_exchange/local_exchanger.cpp [hunk 3]
```diff
@@ -243,6 +245,59 @@ Status PassToOneExchanger::get_block(RuntimeState* state, vectorized::Block* blo
     return Status::OK();
 }
 
+Status LocalMergeSortExchanger::sink(RuntimeState* state, vectorized::Block* in_block, bool eos,
+                                     LocalExchangeSinkLocalState& local_state) {
+    vectorized::Block new_block(in_block->clone_empty());
+    new_block.swap(*in_block);
+    _data_queue[state->per_fragment_instance_idx()].enqueue(std::move(new_block));
```
Comment (Gabriel39) [inline be/src/pipeline/local_exchange/local_exchanger.cpp:252]: Use `local_state._channel_id` to replace `state->per_fragment_instance_idx()`

File: be/src/pipeline/local_exchange/local_exchanger.cpp [hunk 4]
```diff
@@ -243,6 +245,59 @@ Status PassToOneExchanger::get_block(RuntimeState* state, vectorized::Block* blo
...
+}
+
+Status LocalMergeSortExchanger::build_merger(SortSourceOperatorX* sort_source,
+                                             RuntimeState* state) {
+    auto* sort_sink = sort_source->sort_sink();
+    if (!sort_sink) {
+        return Status::InternalError("can not get sort_sink");
+    }
+
+    RETURN_IF_ERROR(sort_sink->build_merger(state, _merger, &_fake_profile));
```
Comment (Gabriel39) [inline be/src/pipeline/local_exchange/local_exchanger.cpp:264]: Move `build_merger` into `SortSourceOperatorX`

Comment (Gabriel39) [reply be/src/pipeline/local_exchange/local_exchanger.cpp:264]: Why we need a _fake_profile?

File: be/src/pipeline/local_exchange/local_exchanger.cpp [hunk 5]
```diff
@@ -243,6 +245,59 @@ Status PassToOneExchanger::get_block(RuntimeState* state, vectorized::Block* blo
...
+            return Status::OK();
+        };
+        _child_block_suppliers.push_back(block_supplier);
+    }
+    return Status::OK();
+}
+
+Status LocalMergeSortExchanger::get_block(RuntimeState* state, vectorized::Block* block, bool* eos,
+                                          LocalExchangeSourceLocalState& local_state) {
+    if (state->per_fragment_instance_idx() != 0) {
```
Comment (Gabriel39) [inline be/src/pipeline/local_exchange/local_exchanger.cpp:287]: local_state._channel_id

File: be/src/pipeline/local_exchange/local_exchanger.cpp [hunk 6]
```diff
@@ -243,6 +245,59 @@ Status PassToOneExchanger::get_block(RuntimeState* state, vectorized::Block* blo
...
+        *eos = true;
+        return Status::OK();
+    }
+    DCHECK(_merger);
+    if (!_has_prepare) {
+        RETURN_IF_ERROR(_merger->prepare(_child_block_suppliers));
+        _merger->set_pipeline_engine_enabled(true);
+        _has_prepare = true;
+    }
+    RETURN_IF_ERROR(_merger->get_next(block, eos));
```
Comment (Gabriel39) [inline be/src/pipeline/local_exchange/local_exchanger.cpp:297]: `block` should be push to free blocks for mem reusing

File: be/src/pipeline/exec/exchange_sink_operator.cpp [hunk 7]
```diff
@@ -664,4 +666,14 @@ Status ExchangeSinkLocalState::close(RuntimeState* state, Status exec_status) {
     return Base::close(state, exec_status);
 }
 
+DataDistribution ExchangeSinkOperatorX::required_data_distribution() const {
+    if (_child_x) {
+        if (auto sort_source = std::dynamic_pointer_cast<SortSourceOperatorX>(_child_x);
```
Comment (Gabriel39) [inline be/src/pipeline/exec/exchange_sink_operator.cpp:671]: Also need a session variable to control

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run performance

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run feut

File: be/src/pipeline/exec/sort_source_operator.cpp [hunk 8]
```diff
@@ -28,7 +29,33 @@ SortLocalState::SortLocalState(RuntimeState* state, OperatorXBase* parent)
...
+    RETURN_IF_ERROR(Base::init(tnode, state));
+    RETURN_IF_ERROR(_vsort_exec_exprs.init(tnode.sort_node.sort_info, _pool));
+    _is_asc_order = tnode.sort_node.sort_info.is_asc_order;
+    _nulls_first = tnode.sort_node.sort_info.nulls_first;
+    return Status::OK();
+}
+
+Status SortSourceOperatorX::prepare(RuntimeState* state) {
+    RETURN_IF_ERROR(Base::prepare(state));
+    if (_child_x) {
```
Comment (Gabriel39) [inline be/src/pipeline/exec/sort_source_operator.cpp:46]: `_child_x` never should be nullptr.

File: be/src/pipeline/exec/sort_source_operator.cpp [hunk 9]
```diff
@@ -28,7 +29,33 @@ SortLocalState::SortLocalState(RuntimeState* state, OperatorXBase* parent)
...
+        RETURN_IF_ERROR(_vsort_exec_exprs.prepare(state, _child_x->row_desc(), _row_descriptor));
+    }
+    return Status::OK();
+}
+
+Status SortSourceOperatorX::open(RuntimeState* state) {
+    RETURN_IF_ERROR(Base::open(state));
+    if (_child_x) {
+        RETURN_IF_ERROR(_vsort_exec_exprs.open(state));
```
Comment (Gabriel39) [inline be/src/pipeline/exec/sort_source_operator.cpp:55]: `_vsort_exec_exprs` should be cloned in local state 's `init`

File: be/src/pipeline/local_exchange/local_exchanger.cpp [hunk 10]
```diff
@@ -261,6 +264,66 @@ Status PassToOneExchanger::get_block(RuntimeState* state, vectorized::Block* blo
...
 
+Status LocalMergeSortExchanger::sink(RuntimeState* state, vectorized::Block* in_block, bool eos,
+                                     LocalExchangeSinkLocalState& local_state) {
+    vectorized::Block new_block;
+    if (!_free_blocks.try_dequeue(new_block)) {
+        new_block = {in_block->clone_empty()};
+    }
+    new_block.swap(*in_block);
+    DCHECK_LE(local_state._channel_id, _data_queue.size());
+    _data_queue[local_state._channel_id].enqueue(std::move(new_block));
```
Comment (Gabriel39) [inline be/src/pipeline/local_exchange/local_exchanger.cpp:275]: We should control memory in _data_queue by sink dependencies

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Review by Gabriel39 — DISMISSED** [review]

**Comment by Mryange** [comment]
run buildall

**Review by Gabriel39 — APPROVED** [review]

**Review by BiteTheDDDDt — APPROVED** [review]

File: be/src/pipeline/local_exchange/local_exchanger.cpp [hunk 11]
```diff
@@ -261,6 +264,98 @@ Status PassToOneExchanger::get_block(RuntimeState* state, vectorized::Block* blo
...
+    add_mem_usage(local_state, new_block.allocated_bytes());
+    local_state._shared_state->set_ready_to_read(0);
```
Comment (HappenLee) [inline be/src/pipeline/local_exchange/local_exchanger.cpp:277]: the logic is error, not add one block can read

**Pull request merged**